// @redux-devtools/utils/lib/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  generateId: true,
  getMethods: true,
  getActionsArray: true,
  evalAction: true,
  evalMethod: true,
  stringify: true,
  getSeralizeParameter: true,
  getStackTrace: true
};
exports.evalAction = evalAction;
exports.evalMethod = evalMethod;
exports.generateId = generateId;
exports.getActionsArray = getActionsArray;
exports.getMethods = getMethods;
exports.getSeralizeParameter = getSeralizeParameter;
exports.getStackTrace = getStackTrace;
exports.stringify = stringify;
var _getParams = _interopRequireDefault(require("get-params"));
var _jsan = _interopRequireDefault(require("jsan"));
var _nonSecure = require("nanoid/non-secure");
var _serialize = require("@redux-devtools/serialize");
var _catchErrors = require("./catchErrors");
Object.keys(_catchErrors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _catchErrors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _catchErrors[key];
    }
  });
});
var _filters = require("./filters");
Object.keys(_filters).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _filters[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _filters[key];
    }
  });
});
var _importState = require("./importState");
Object.keys(_importState).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _importState[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _importState[key];
    }
  });
});
function generateId(id) {
  return id || (0, _nonSecure.nanoid)(7);
}
function flatTree(obj) {
  let namespace = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
  let functions = [];
  Object.keys(obj).forEach(key => {
    const prop = obj[key];
    if (typeof prop === 'function') {
      functions.push({
        name: namespace + (key || prop.name || 'anonymous'),
        func: prop,
        args: (0, _getParams.default)(prop)
      });
    } else if (typeof prop === 'object') {
      functions = functions.concat(flatTree(prop, namespace + key + '.'));
    }
  });
  return functions;
}
function getMethods(obj) {
  if (typeof obj !== 'object') return undefined;
  let functions;
  let m;
  if (obj.__proto__) m = obj.__proto__.__proto__;
  if (!m) m = obj;
  Object.getOwnPropertyNames(m).forEach(key => {
    const propDescriptor = Object.getOwnPropertyDescriptor(m, key);
    if (!propDescriptor || 'get' in propDescriptor || 'set' in propDescriptor) return;
    const prop = m[key];
    if (typeof prop === 'function' && key !== 'constructor') {
      if (!functions) functions = [];
      functions.push({
        name: key || prop.name || 'anonymous',
        args: (0, _getParams.default)(prop)
      });
    }
  });
  return functions;
}
function getActionsArray(actionCreators) {
  if (Array.isArray(actionCreators)) return actionCreators;
  return flatTree(actionCreators);
}
const interpretArg = arg =>
// eslint-disable-next-line @typescript-eslint/no-implied-eval
new Function('return ' + arg)();
function evalArgs(inArgs, restArgs) {
  const args = inArgs.map(interpretArg);
  if (!restArgs) return args;
  const rest = interpretArg(restArgs);
  if (Array.isArray(rest)) return args.concat(...rest);
  throw new Error('rest must be an array');
}
function evalAction(action, actionCreators) {
  if (typeof action === 'string') {
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    return new Function('return ' + action)();
  }
  const actionCreator = actionCreators[action.selected].func;
  const args = evalArgs(action.args, action.rest);
  return actionCreator(...args);
}
function evalMethod(action, obj) {
  if (typeof action === 'string') {
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    return new Function('return ' + action).call(obj);
  }
  const args = evalArgs(action.args, action.rest);
  // eslint-disable-next-line @typescript-eslint/no-implied-eval
  return new Function('args', `return this.${action.name}(args)`).apply(obj, args);
}
/* eslint-enable */

function tryCatchStringify(obj) {
  try {
    return JSON.stringify(obj);
  } catch (err) {
    /* eslint-disable no-console */
    if (process.env.NODE_ENV !== 'production') console.log('Failed to stringify', err);
    /* eslint-enable no-console */
    return _jsan.default.stringify(obj, null, null, {
      circular: '[CIRCULAR]'
    });
  }
}
function stringify(obj, serialize) {
  if (typeof serialize === 'undefined') {
    return tryCatchStringify(obj);
  }
  if (serialize === true) {
    return _jsan.default.stringify(obj, function (key, value) {
      if (value && typeof value.toJS === 'function') return value.toJS();
      return value;
    }, null, true);
  }
  return _jsan.default.stringify(obj, serialize.replacer, null, serialize.options);
}
function getSeralizeParameter(config, param) {
  const serialize = config.serialize;
  if (serialize) {
    if (serialize === true) return {
      options: true
    };
    if (serialize.immutable) {
      return {
        replacer: (0, _serialize.immutableSerialize)(serialize.immutable, serialize.refs).replacer,
        options: serialize.options || true
      };
    }
    if (!serialize.replacer) return {
      options: serialize.options
    };
    return {
      replacer: serialize.replacer,
      options: serialize.options || true
    };
  }
  const value = config[param];
  if (typeof value === 'undefined') return undefined;
  // eslint-disable-next-line no-console
  console.warn(`\`${param}\` parameter for Redux DevTools Extension is deprecated. Use \`serialize\` parameter instead:` + ' https://github.com/zalmoxisus/redux-devtools-extension/releases/tag/v2.12.1');
  return value;
}
function getStackTrace(
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
config,
// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
toExcludeFromTrace) {
  if (!config.trace) return undefined;
  if (typeof config.trace === 'function') return config.trace();
  let stack;
  let extraFrames = 0;
  let prevStackTraceLimit;
  const traceLimit = config.traceLimit;
  const error = Error();
  if (Error.captureStackTrace) {
    if (Error.stackTraceLimit < traceLimit) {
      prevStackTraceLimit = Error.stackTraceLimit;
      Error.stackTraceLimit = traceLimit;
    }
    Error.captureStackTrace(error, toExcludeFromTrace);
  } else {
    extraFrames = 3;
  }
  stack = error.stack;
  if (prevStackTraceLimit) Error.stackTraceLimit = prevStackTraceLimit;
  if (extraFrames || typeof Error.stackTraceLimit !== 'number' || Error.stackTraceLimit > traceLimit) {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    const frames = stack.split('\n');
    if (frames.length > traceLimit) {
      stack = frames.slice(0, traceLimit + extraFrames + (frames[0] === 'Error' ? 1 : 0)).join('\n');
    }
  }
  return stack;
}
